auth/auth_test.go

```diff
@@ -1240,3 +1240,933 @@ func TestAuthenticateUntrustedJWT(t *testing.T) {
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+	assert.Equal(t, 0, len(fooPrincipal.ChannelHistory()))
+
+	// Get Principals / Rebuild Seq 40
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user can see ch1 (via role)
+	// Verify history
+	assert.ElementsMatch(t, []string{"!", "ch1"}, fooPrincipal.Channels().AllChannels())
+	assert.True(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	assert.Equal(t, 0, len(aliceUserPrincipal.RoleHistory()))
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+	assert.Equal(t, 0, len(fooPrincipal.ChannelHistory()))
+
+	testMockComputer.removeRole(t, auth, aliceUserPrincipal, "alice", "foo", 45)
+	testMockComputer.removeRoleChannel(t, auth, fooPrincipal, "foo", "ch1", 55)
+
+	testMockComputer.addRole(t, auth, aliceUserPrincipal, "alice", "foo", 65)
+	testMockComputer.addRoleChannels(t, auth, fooPrincipal, "foo", "ch1", 75)
+
+	// Get Principals / Rebuild Seq 80
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user can see ch1 (via role)
+	// Verify history
+	assert.ElementsMatch(t, []string{"!", "ch1"}, fooPrincipal.Channels().AllChannels())
+	assert.True(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	assert.Equal(t, 0, len(aliceUserPrincipal.RoleHistory()))
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+	assert.Equal(t, 0, len(fooPrincipal.ChannelHistory()))
+
+	testMockComputer.removeRoleChannel(t, auth, fooPrincipal, "foo", "ch1", 85)
+	testMockComputer.removeRole(t, auth, aliceUserPrincipal, "alice", "foo", 95)
+
+	// Get Principals / Rebuild Seq 110
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user cannot see ch1 (via role)
+	// Verify history
+	assert.False(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	userRoleHistory, ok := aliceUserPrincipal.RoleHistory()["foo"]
+	require.True(t, ok)
+	assert.Equal(t, GrantHistorySequencePair{Seq: 65, EndSeq: 95}, userRoleHistory.Entries[0])
+	channelHistory, ok := fooPrincipal.ChannelHistory()["ch1"]
+	require.True(t, ok)
+	assert.Equal(t, GrantHistorySequencePair{Seq: 75, EndSeq: 85}, channelHistory.Entries[0])
+}
+
+// Scenario 2
+// Get Doc
+// Revoke doc role
+// Re-grant
+// Re-revoke via both role and channel
+func TestRevocationScenario2(t *testing.T) {
+	testBucket := base.GetTestBucket(t)
+	defer testBucket.Close()
+
+	testMockComputer := mockComputerV2{
+		roles:        map[string]ch.TimedSet{},
+		channels:     map[string]ch.TimedSet{},
+		roleChannels: map[string]ch.TimedSet{},
+	}
+
+	auth := NewAuthenticator(testBucket, &testMockComputer)
+	aliceUserPrincipal, fooPrincipal := initializeScenario(t, auth)
+
+	testMockComputer.addRoleChannels(t, auth, fooPrincipal, "foo", "ch1", 5)
+	testMockComputer.addRole(t, auth, aliceUserPrincipal, "alice", "foo", 20)
+
+	// Get Principals / Rebuild Seq 25
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user can see ch1 (via role)
```

Comment:
These tests seem to be doing 'ensure user can see channel' as a surrogate for what's shown as changes requests in the spreadsheet, but I don't think that covers the key question, which is 'would a changes request at this sequence trigger a revocation'.  

I don't think you need to implement the entire changes handling to do that, but I feel like it does make sense to have a test function that performs the same evaluation (of role history, user channel history and user role history) for that.  Is the idea that you'd enhance these tests when you implement the changes-side work?

Reply:
The can see channel is a part of it but not entirely. The point in these tests are to verify that the history is built correctly and not necessarily that we would revoke. The revoke tests would come later. Once the changes-side work is complete there will be similar tests doing the same scenarios but in full. With actual REST calls to create the user / docs and issue changes request to ensure the document is revoked.

Reply:
So is the reason the tests are doing 'ensure user can see channel' just to perform auth, so that the channel history gets resolved?  Probably best to state that explicitly.

Reply:
No. GetPrincipals does the auth. Can probably drop the 'ensure user can see channel' really

Reply:
Should be addressed as per discussion